massa-event-cache/src/event_cache.rs

```diff
@@ -0,0 +1,1730 @@
+// std
+use std::cmp::max;
+use std::collections::{BTreeMap, BTreeSet};
+use std::path::Path;
+// third-party
+use num_enum::IntoPrimitive;
+use rocksdb::{IteratorMode, Options, WriteBatch, DB};
+use tracing::{debug, warn};
+// internal
+use crate::rocksdb_operator::counter_merge;
+use crate::ser_deser::{
+    SCOutputEventDeserializer, SCOutputEventDeserializerArgs, SCOutputEventSerializer,
+};
+use massa_models::address::Address;
+use massa_models::error::ModelsError;
+use massa_models::execution::EventFilter;
+use massa_models::operation::{OperationId, OperationIdSerializer};
+use massa_models::output_event::SCOutputEvent;
+use massa_models::slot::Slot;
+use massa_serialization::{DeserializeError, Deserializer, Serializer};
+
+const OPEN_ERROR: &str = "critical: rocksdb open operation failed";
+// const COUNTER_INIT_ERROR: &str = "critical: cannot init rocksdb counters";
+const DESTROY_ERROR: &str = "critical: rocksdb delete operation failed";
+const CRUD_ERROR: &str = "critical: rocksdb crud operation failed";
+const EVENT_DESER_ERROR: &str = "critical: event deserialization failed";
+const OPERATION_ID_DESER_ERROR: &str = "critical: deserialization failed for op id in rocksdb";
+const COUNTER_ERROR: &str = "critical: cannot get counter";
+const COUNTER_KEY_CREATION_ERROR: &str = "critical: cannot create counter key";
+
+#[allow(dead_code)]
+/// Prefix u8 used to identify rocksdb keys
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, IntoPrimitive)]
+#[repr(u8)]
+enum KeyIndent {
+    Counter = 0,
```

Comment:
The fact that Counter is an indent makes the behavior very confusing to read.

I would either have 
```
enum KeyIndent {
    EventById,
    EventByIdCounter,
    EventByEmitterAddress,
    EventByEmitterAddressCounter,
    ...
}

// here the format of the EventByEmitterAddress index would be:
//   [KeyIndent::EventByEmitterAddressCounter][len(index_value)][index_value] -> (value = number of items)
//   [KeyIndent::EventByEmitterAddress][len(index_value)][index_value][event_id] -> (no value)
```

or clearly separate the two stages:
```
enum KeyIndent {
    EventById,
    EventByEmitterAddress,
    ...
}

enum KeySubtype {
    Counter,
    Item
}

// here the format of the EventByEmitterAddress index would be:
//   [KeyIndent::EventByIdCounter][len(index_value)][index_value][KeySubtype::Counter] -> (value = number of items)
//   [KeyIndent::EventByIdCounter][len(index_value)][index_value][KeySubtype::Item][event_id] -> (no value)
```

Reply:
In the code, the key format is the following (as documented in DbKeyBuilder::key_from_event method): 

```
        // Db format:
        // * Regular keys:
        //   * Event key: [Event Indent][Slot][Index] -> Event value: Event serialized
        //   * Emitter address key: [Emitter Address Indent][Addr][Addr len][Event key] -> []
        // * Prefix keys:
        //   * Emitter address prefix key: [Emitter Address Indent][Addr][Addr len]
        // * Counter keys:
        //   * Emitter address counter key: [Counter indent][Emitter Address Indent][Addr][Addr len][Event key] -> u64
```

This format has some advantages like:
* Read or Identify the kind/type of key easily (by just looking at the first byte)
* Prefix iteration only return the "regular" key and not the counter key:
  * by using a prefix key for an emitter address, only the emitter address keys are returned and not the emitter address counter key  
* Easily iterate over some keys with just an indent. It's is used in EventCache::snip method and in unit tests (test_counter_0)

 Not sure about the advantages of your solution...

Reply:
The simplified builder helps. OK with the design